app/services/exports/export_request_service.rb

```diff
@@ -1,48 +1,112 @@
       end
     end
 
+    def generate_csv_data
+      csv_data = []
+
+      csv_data << headers
+      requests.each do |request|
+        csv_data << build_row_data(request)
+      end
+
+      csv_data
+    end
+
     private
 
-    attr_reader :requests, :headers
-
-    def rows
-      requests.map do |request|
-        {
-          "Date" => request.created_at.strftime("%m/%d/%Y"),
-          "Requestor" => request.partner.name,
-          "Status" => request.status.humanize
-        }.tap do |row|
-          request.request_items.each do |item_ref|
-            item = grouped_items[item_ref['item_id'].to_i]&.first
-            row[item.name] = item_ref['quantity']
-            headers << item.name unless headers.include?(item.name)
-          end
+    attr_reader :requests
+
+    def headers
+      # Build the headers in the correct order
+      base_headers + item_headers
+    end
+
+    def headers_with_indexes
+      @headers_with_indexes ||= headers.each_with_index.to_h
+    end
+
+    def base_table
+      {
+        "Date" => ->(request) {
+          request.created_at.strftime("%m/%d/%Y")
+        },
+        "Requestor" => ->(request) {
+          request.partner.name
+        },
+        "Status" => ->(request) {
+          request.status.humanize
+        }
+      }
+    end
+
+    def base_headers
+      base_table.keys
+    end
+
+    def item_headers
+      item_names = items.pluck(:name)
+
+      # Adding this to handle cases in which a requested item
+      # has been deleted. Normally this wouldn't be neccessary,
+      # but previous versions of the application would cause
+      # this orphaned data
+      item_names.sort.uniq << deleted_item_column_header
+    end
+
+    def build_row_data(request)
+      row = base_table.values.map { |closure| closure.call(request) }
+
+      row += Array.new(item_headers.size, 0)
+
+      request.request_items.each do |request_item|
+        item_name = fetch_item_name(request_item['item_id']) || deleted_item_column_header
+        item_column_idx = headers_with_indexes[item_name]
+
+        if item_name == deleted_item_column_header
+          # Add to the deleted column for every item that
+          # does not match any existing Item.
+          row[item_column_idx] ||= 0
+          row[item_column_idx] += request_item['quantity']
+        else
+          row[item_column_idx] = request_item['quantity']
         end
       end
+
+      row
     end
 
-    def grouped_items
-      @grouped_items ||= Item.where(id: request_item_ids).group_by(&:id)
+    def fetch_item_name(item_id)
+      @item_name_to_id_map ||= items.inject({}) do |acc, item|
+        acc[item.id] = item.name
+        acc
+      end
+
+      @item_name_to_id_map[item_id]
     end
 
-    def request_item_ids
-      requests.map do |request|
+    def deleted_item_column_header
+      "<DELETED ITEMS>"
+    end
+
+    def items
+      return @items if @items
+
+      item_ids = requests.map do |request|
```

Comment:
nit: map + flatten could be replaced with `flat_map`, but this is fine too

Reply:
Nice! I changed to use that. I didn't know about `flat_map`!